Manage Transitive Dependencies for Static Libraries

ollieatkinson opened the pull request:
Resolves https://github.com/tuist/tuist/issues/60

### Short description 📝 & Solution 📦

Static Libraries were not yet fully supported, the dependency graph would include the static libraries in each dependency producing duplicate symbols.

- Include all transitive dependencies in top level bundles when using static libraries
- Create "Fake Swift Dependencies" to allow Xcode to build the static libraries in the correct order 
- WIP

### Implementation 👩‍💻👨‍💻

- Implemented DFS in `Graph.linkableDependencies` to search for all static libraries, this will include them if the top level bundle is an app, ui or unit test target.
- Continue to link dynamic libraries and frameworks the same way, this needs further testing
- Ignore linking any static libraries to static libraries (need to check if we would ever actually want this, see L142)
- Implement `Graph.staticLibraryDependencies` to resolve a one-level-deep dependency for the use of generating the fake dependencies build phase 
- Implement `LinkGenerator.generateFakeSwiftDependenciesForStaticLibraries` to create a copy files phase which act's as a build file dependency resolution.

This is a WIP so any guidance would be appreciated.

ollieatkinson commented:
Attached is the example project I have been using to validate these changes
[Example.zip](https://github.com/tuist/tuist/files/2653336/Example.zip)

pepibumur commented:
> Attached is the example project I have been using to validate these changes
Example.zip

I created [this repository](https://github.com/tuist/example-project) where we can push a project for testing purposes.

pepibumur commented on Sources/TuistKit/Loader/Graph.swift:
```diff
@@ -201,3 +254,27 @@ class Graph: Graphing {
     }
 }
+
+public struct Stack<T> {
```
I'd extract this into its own file. Maybe in [this directory](https://github.com/tuist/tuist/tree/master/Sources/TuistKit/Utils)

pepibumur replied:
I'd also add a few unit tests.

ollieatkinson replied:
Moved the code to `Utils/`

pepibumur commented on Sources/TuistKit/Loader/Graph.swift:
```diff
@@ -120,20 +136,61 @@ class Graph: Graphing {
-                let xcodeProduct = targetNode.target.product.xcodeValue
-                guard let `extension` = xcodeProduct.fileExtension else {
-                    throw GraphError.unsupportedFileExtension(xcodeProduct.rawValue)
+            var stack = Stack<TargetNode>()
```
The traversing logic is worth extracting into `Graph`.  I can see it being used like:

```swift
graph.traverse(from: node) { node in
  // Do whatever you want with the node
}
```

What do you think @ollieatkinson ?

ollieatkinson replied:
I implemented the following and exposed their interface on `Graphing`

```
    func findAll<T: GraphNode>(path: AbsolutePath, test: (T) -> Bool) -> Set<T>
    func findAll<T: GraphNode>(path: AbsolutePath, name: String, test: (T) -> Bool) -> Set<T>
```

pepibumur commented on Sources/TuistKit/Loader/Graph.swift:
```diff
@@ -109,6 +110,21 @@ class Graph: Graphing {
+        
+    }
 
     func linkableDependencies(path: AbsolutePath, name: String) throws -> [DependencyReference] {
```
I'd add a test to check if we are traversing the graph properly and returning the expected list of static libraries.

ollieatkinson replied:
Implemented `func test_trasistiveDependenciesForStaticLibraries()`

pepibumur commented on Sources/TuistKit/Generator/LinkGenerator.swift:
```diff
@@ -203,4 +215,43 @@ final class LinkGenerator: LinkGenerating {
+        let buildPhase = PBXCopyFilesBuildPhase(
+            dstPath: nil,
+            dstSubfolderSpec: .productsDirectory,
+            name: "[tuist] Fake Swift Dependencies",
```
I'd avoid having references to `tuist` from the generated projects and files. What about calling the phase:

```
Transitive static dependencies
```

ollieatkinson replied:
I'm not sure "Transitive static dependencies" makes sense for the build phase name since they are actually the direct dependents of the library (not every single dependency).

maybe just "Dependencies" makes more sense here

ollieatkinson replied:
Renamed to `Dependencies`

pepibumur commented on Sources/TuistKit/Generator/LinkGenerator.swift:
```diff
@@ -203,4 +214,38 @@ final class LinkGenerator: LinkGenerating {
         }
     }
+
+    func generateFakeSwiftDependenciesForStaticLibraries(dependencies: [DependencyReference],
```
What about renaming this to `generateDependenciesBuildPhase`?

pepibumur commented on Sources/TuistKit/Generator/LinkGenerator.swift:
```diff
@@ -203,4 +214,38 @@ final class LinkGenerator: LinkGenerating {
+                                                         fileElements: ProjectFileElements) throws {
+        var files: [PBXBuildFile] = [ ]
+
+        for case let .product(name) in dependencies {
```
This method looks much cleaner now. Great job 👍

pepibumur commented on Sources/TuistKit/Generator/ProjectFileElements.swift:
```diff
@@ -54,7 +54,14 @@ class ProjectFileElements {
+            return true
+        }
+        
+        let dependencies = graph.findAll(path: project.path, test: alwaysTrue)
```
Nitpick: What about inlining the closure:

```swift
let dependencies = graph.findAll(path: project.path, test: { _ in true })
```

Also we can use `@autoclosure` in the `test` argument and write it evens shorter:

```swift
let dependencies = graph.findAll(path: project.path, test: true)
```

_(Not sure if the code above would work though)_

ollieatkinson replied:
I hadn't thought about that, I like the implementation - let's go for it!

ollieatkinson replied:
I opted for just defining another function which defaults the test to always true

```
func findAll<T: GraphNode>(path: AbsolutePath) -> Set<T>
```

pepibumur commented on Sources/TuistKit/Loader/Graph.swift:
```diff
@@ -46,12 +47,18 @@ protocol Graphing: AnyObject {
     func targetDependencies(path: AbsolutePath, name: String) -> [String]
+    func staticLibraryDependencies(path: AbsolutePath, name: String) -> [DependencyReference]
+    
+    // MARK:- DFS
```
What does DFS stand for? Would you mind expanding it?

ollieatkinson replied:
Depth First Search, I'll add some header documentation to what it is and why we need it.

pepibumur commented on Sources/TuistKit/Loader/Graph.swift:
```diff
@@ -77,78 +84,91 @@ class Graph: Graphing {
+        case .app, .unitTests, .uiTests:
+
+            let staticLibraries = findAll(targetNode: targetNode, test: isStaticLibrary(targetNode:))
+                .lazy
```
TIL about this operator. Nice idea 💡

pepibumur commented on Sources/TuistKit/Loader/Graph.swift:
```diff
@@ -77,78 +84,91 @@ class Graph: Graphing {
-                return DependencyReference.product("\(targetNode.target.name).\(`extension`)")
-        }))
+
+        let precompiledLibrariesAndFrameworks = targetNode.precompiledDependencies
```
Same here, this is very readable and easy to reason about 👏

pepibumur commented on Sources/TuistKit/Loader/Graph.swift:
```diff
@@ -164,40 +184,167 @@ class Graph: Graphing {
+
+// MARK:- TargetNode helper computed properties, provide lazy arrays by default.
+
+extension TargetNode {
```
Good idea 👏

pepibumur commented on Sources/TuistKit/Utils/Functions.swift:
```diff
@@ -0,0 +1,36 @@
+/// Produces a getter function for a given key path. Useful for composing property access with functions.
```
```swift
import Foundation
```

ollieatkinson replied:
see 
https://github.com/tuist/tuist/pull/168#discussion_r239753274

pepibumur commented on Sources/TuistKit/Utils/Sequence+KeyPath.swift:
```diff
@@ -0,0 +1,23 @@
+extension Sequence {
```
```swift
import Foundation
```

ollieatkinson replied:
Technically these don't need Foundation as they only use code in the Stdlib

pepibumur commented on Sources/TuistKit/Utils/Functions.swift:
```diff
@@ -0,0 +1,36 @@
+/// Produces a getter function for a given key path. Useful for composing property access with functions.
```
What about moving this extension to the [`Extensions` directory](https://github.com/tuist/tuist/tree/master/Sources/TuistCore/Extensions) of `TuistCore`?

pepibumur commented on Sources/TuistKit/Utils/Sequence+KeyPath.swift:
```diff
@@ -0,0 +1,23 @@
+extension Sequence {
```
Same comment here regarding moving the extension to `TuistCore` and making it public.

pepibumur commented on Sources/TuistKit/Utils/Stack.swift:
```diff
@@ -0,0 +1,21 @@
+/// Implements a Stack - helper class for push/pop that uses an array internally.
+public struct Stack<T> {
```
I'd also move this one.

pepibumur commented on Tests/TuistKitTests/Utils/StackTests.swift:
```diff
@@ -0,0 +1,73 @@
+//
```
Don't forget to remove the header comments.

pepibumur commented on Sources/TuistCore/Utils/Functions.swift:
```diff
@@ -0,0 +1,36 @@
+/// Produces a getter function for a given key path. Useful for composing property access with functions.
```
```swift
import Foundation
```

ollieatkinson replied:
I mentioned this before, these don't actually require `Foundation` as they only the standard lib - do you still want it here even though it's redundant?

pepibumur commented on Sources/TuistCore/Utils/Stack.swift:
```diff
@@ -0,0 +1,24 @@
+/// Implements a Stack - helper class for push/pop that uses an array internally.
```
```swift
import Foundation
```

pepibumur commented on Sources/TuistCore/Extensions/Sequence+KeyPath.swift:
```diff
@@ -0,0 +1,23 @@
+extension Sequence {
```
```swift
import Foundation
```

ollieatkinson commented:
I've rebased and updated the branch against the changes in master.
I've created some examples over here which we can use to validate changes over time: https://github.com/tuist/example-project/pull/1

pepibumur commented:
The PR looks good to me. There's only one more thing, what do you think about moving those example projects into this repository? We could set up a battery of integration tests that run on those projects. We can do that in a follow-up PR.

ollieatkinson commented:
@pepibumur Sure, we can do that, would the simple case be executing the unit tests for each of those modules?

pepibumur commented:
For each of those projects, I'd first run a test that makes sure that the targets in the project compile. If the compilation succeeds, we can run some assertions on the build products. For example, if we expect some dynamic frameworks to be copied into the build product, we can verify if the frameworks exist in the product frameworks directory. 

If you there's anything else worth asserting, feel free to add those tests as well. By having those tests here we can make sure no one breaks the linking logic, and that the outputs from Xcode have the format that we are expecting.

I'd write those tests with Cucumber (in Ruby), have you used that framework before? Let me know if you need some guidance to write them, I'm happy to help.

Feel free to do it in a follow-up PR so that that we don't get this PR larger. You should have persmissions to merge this PR, so feel free to go ahead with the merge.

ollieatkinson commented:
I've captured the work inside a ticket https://github.com/tuist/tuist/issues/198, I am not able to merge the branch - is this because I forked the repo and the automated check for the tests has not passed?

pepibumur commented:
> I've captured the work inside a ticket #198, I am not able to merge the branch - is this because I forked the repo and the automated check for the tests has not passed?

Thanks for opening the issue.
Yeah, I think it must be because the PR is opened from a fork. I'll merge it then.
